src/libtiled/mapwriter.cpp

```diff
@@ -648,45 +650,83 @@ void MapWriterPrivate::writeTileLayer(QXmlStreamWriter &w,
         QRect bounds = tileLayer.bounds().translated(-tileLayer.position());
         startX = bounds.left();
         startY = bounds.top();
-        endX = bounds.right();
-        endY = bounds.bottom();
+        endX = bounds.right() + 1;
+        endY = bounds.bottom() + 1;
+
+        int chunkStartX = startX;
+        int chunkStartY = startY;
+        int resetX = chunkStartX;
+
+        startX = (startX < 0) ? (startX + 1) / CHUNK_SIZE - 1 : startX / CHUNK_SIZE;
+        startY = (startY < 0) ? (startY + 1) / CHUNK_SIZE - 1 : startY / CHUNK_SIZE;
+        endX = (endX < 0) ? (endX + 1) / CHUNK_SIZE - 1 : endX / CHUNK_SIZE;
+        endY = (endY < 0) ? (endY + 1) / CHUNK_SIZE - 1 : endY / CHUNK_SIZE;
+
+        for (int y = startY; y < endY; ++y) {
+            for (int x = startX; x < endX; ++x) {
+                if (tileLayer.findChunk(chunkStartX, chunkStartY)) {
+                    w.writeStartElement(QLatin1String("chunk"));
+                    w.writeAttribute(QLatin1String("x"), QString::number(chunkStartX));
+                    w.writeAttribute(QLatin1String("y"), QString::number(chunkStartY));
+                    w.writeAttribute(QLatin1String("width"), QString::number(CHUNK_SIZE));
+                    w.writeAttribute(QLatin1String("height"), QString::number(CHUNK_SIZE));
+
+                    writeTileLayerData(w, tileLayer, QRect(chunkStartX, chunkStartY, CHUNK_SIZE, CHUNK_SIZE));
+
+                    w.writeEndElement(); // </chunk>
+                }
+
+                chunkStartX += CHUNK_SIZE;
+            }
+
+            chunkStartX = resetX;
+            chunkStartY += CHUNK_SIZE;
+        }
+    } else {
+        writeTileLayerData(w, tileLayer, QRect(0, 0, endX + 1, endY + 1));
```

Comment:
Judging by the code above, `endX + 1` and `endY + 1` can't possibly be the right values here, and should rather be `tileLayer.width()` and `tileLayer.height()`.

Reply:
```
int startX = 0;
int startY = 0;
int endX = tileLayer.width() - 1;
int endY = tileLayer.height() - 1;
```

`endX + 1` and `endY + 1` are same as `tileLayer.width()` and `tileLayer.height()`.

Reply:
Ah, that's a little confusing. Please just use `tileLayer.width()` and `tileLayer.height()` here and move those variables local to the other branch.